core/src/main/scala/org/apache/spark/storage/PushBasedFetchHelper.scala

```diff
@@ -0,0 +1,336 @@
+   * given merged blocks, split them into remote and local blocks, and process them
+   * accordingly.
+   * The fallback happens when:
+   * 1. There is an exception while creating shuffle block chunk from local merged shuffle block.
+   *    See fetchLocalBlock.
+   * 2. There is a failure when fetching remote shuffle block chunks.
+   * 3. There is a failure when processing SuccessFetchResult which is for a shuffle chunk
+   *    (local or remote).
+   *
+   * @return number of blocks processed
+   */
+  def initiateFallbackBlockFetchForMergedBlock(
+      blockId: BlockId,
+      address: BlockManagerId): Int = {
+    assert(blockId.isInstanceOf[ShuffleBlockId] || blockId.isInstanceOf[ShuffleBlockChunkId])
+    logWarning(s"Falling back to fetch the original unmerged blocks for merged block $blockId")
+    // Increase the blocks processed since we will process another block in the next iteration of
+    // the while loop in ShuffleBlockFetcherIterator.next().
+    var blocksProcessed = 1
+    val fallbackBlocksByAddr: Iterator[(BlockManagerId, Seq[(BlockId, Long, Int)])] =
+      blockId match {
+        case shuffleBlockId: ShuffleBlockId =>
+          mapOutputTracker.getMapSizesForMergeResult(
+            shuffleBlockId.shuffleId, shuffleBlockId.reduceId)
+        case _ =>
+          val shuffleChunkId = blockId.asInstanceOf[ShuffleBlockChunkId]
+          val chunkBitmap: RoaringBitmap = chunksMetaMap.remove(shuffleChunkId)
+          assert(chunkBitmap != null)
+          // When there is a failure to fetch a remote merged shuffle block chunk, then we try to
+          // fallback not only for that particular remote shuffle block chunk but also for all the
+          // pending block chunks that belong to the same host. The reason for doing so is that it
+          // is very likely that the subsequent requests for merged block chunks from this host will
+          // fail as well. Since, push-based shuffle is best effort and we try not to increase the
+          // delay of the fetches, we immediately fallback for all the pending shuffle chunks in the
+          // fetchRequests queue.
+          if (isMergedBlockAddressRemote(address)) {
+            // Fallback for all the pending fetch requests
+            val pendingShuffleChunks = iterator.removePendingChunks(shuffleChunkId, address)
+            if (pendingShuffleChunks.nonEmpty) {
+              pendingShuffleChunks.foreach { pendingBlockId =>
+                logInfo(s"Falling back immediately for merged block $pendingBlockId")
+                val bitmapOfPendingChunk: RoaringBitmap = chunksMetaMap.remove(pendingBlockId)
+                assert(bitmapOfPendingChunk != null)
+                chunkBitmap.or(bitmapOfPendingChunk)
+              }
+              // These blocks were added to numBlocksToFetch so we increment numBlocksProcessed
+              blocksProcessed += pendingShuffleChunks.size
```

Comment:
Do you mean in the iterator we decrease `numBlocksToFetch` instead of increasing `numBlocksProcessed`?

Reply:
I am making the change to decrease `numBlocksToFetch` instead of increasing `numBlocksProcessed`

Reply:
Just simple decreasing `numBlocksToFetch` creates issues because the  `iterator.next()` increases `numBlocksProcessed` right at the beginning for all kind of results. So, now if we just decrease `numBlocksToFetch`, these 2 get out of sync.

Reply:
yes. IIUC, it's equal.

Reply:
I checked all cases where call `initiateFallbackBlockFetchForMergedBlock`. It looks like all cases set `result` to `null` after the call of `initiateFallbackBlockFetchForMergedBlock`. So in these cases, they don't increase `numBlocksProcessed` repeatedly, right?

Reply:
Yeah they won't increased `numBlocksProcessed` repeatedly. Some of the tests that I have added are failing after making this change and the one for https://github.com/apache/spark/pull/32140#discussion_r653103522
The iterator gets exhausted prematurely. I need to check what's the problem.
I will work on it to make this change though.

Reply:
I have made this change as well. `initiateFallbackFetchForPushMergedBlock` is now decrementing `numBlocksToFetch`